core/store/src/flat/storage.rs

```diff
@@ -98,6 +98,12 @@ impl FlatStorageInner {
 }
 
 impl FlatStorage {
+    /// Expected limits for in-memory stored changes, under which flat storage must keep working.
+    /// If they are exceeded, warnings are displayed. Flat storage still will work, but its
+    /// performance will slow down, and eventually it can cause OOM error.
+    const CACHED_CHANGES_LIMIT: usize = 50;
```

Comment:
Why do we want to track this separately? As far as I understand there is no risk of having more than 50 small deltas cached and producing `warn` in this case is a false-alarm, at least in the context in flat storage. I suggest removing that and only tracking the total size of cached changes.

Reply:
As far as I remember, there is a risk of node slowdown if there are too many deltas, even if they are small, so it is also worth a warning.

Reply:
Please elaborate on what those risks are. In my understanding the bottleneck is hash map access, but then this value should be much larger: I expect main memory access to be `~100ns` (which is a pessimistic case with processor cache miss). Even assuming we need 2 memory accesses for each hash map access and acceptable flat storage latency is `100us` then we can still support 500 deltas without major performance degradation.
In general I'm against unnecessary tracking/warn because it introduces maintenance overhead for the future. If you insist on keeping that then please make sure to write an elaborate comment describing calculation behind the value as well as what could happen when we exceed it. For example current comment `eventually it can cause OOM error.` behind `CACHED_CHANGES_LIMIT` is very misleading, because that is only applicable to `CACHED_CHANGES_SIZE_LIMIT`

Reply:
ah, looks like @jakmeier has just proven me wrong: https://github.com/near/nearcore/issues/8006#issuecomment-1473621334 and we can reasonable support around only around 100 deltas 😞

Reply:
Tbh I didn't know exact numbers, but now I can happily add them to comments

Reply:
Let's keep it at `100` and add a link to Jakob's table in the comment